fails to build with missing file

When building on my Trusty VM with `./src/ament/ament_tools/scripts/ament.py build ./src --build-tests`:

```
Process package 'uncrustify' with context:
--------------------------------------------------------------------------------
 source_space => /home/william/ros2_ws/src/ament/uncrustify
  build_space => /home/william/ros2_ws/build/uncrustify
install_space => /home/william/ros2_ws/install
   make_flags => -j4, -l4
  build_tests => True
--------------------------------------------------------------------------------
+++ Building 'uncrustify'
Running cmake because arguments have changed.
==> '. /home/william/ros2_ws/build/uncrustify/ament_cmake__build.sh && /usr/bin/cmake /home/william/ros2_ws/src/ament/uncrustify -DCMAKE_INSTALL_PREFIX=/home/william/ros2_ws/install -DAMENT_ENABLE_TESTING=1' in '/home/william/ros2_ws/build/uncrustify'
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    AMENT_ENABLE_TESTING


-- Build files have been written to: /home/william/ros2_ws/build/uncrustify
==> '. /home/william/ros2_ws/build/uncrustify/ament_cmake__build.sh && /usr/bin/make -j4 -l4' in '/home/william/ros2_ws/build/uncrustify'
Scanning dependencies of target external_uncrustify
[ 12%] Creating directories for 'external_uncrustify'
[ 25%] No download step for 'external_uncrustify'
[ 50%] [ 50%] No patch step for 'external_uncrustify'
No update step for 'external_uncrustify'
[ 62%] Performing configure step for 'external_uncrustify'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for style of include used by make... GNU
checking dependency style of g++... none
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for inttypes.h... (cached) yes
checking for stdint.h... (cached) yes
checking utime.h usability... yes
checking utime.h presence... yes
checking for utime.h... yes
checking for stdbool.h that conforms to C99... yes
checking for _Bool... yes
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for working memcmp... yes
checking for memset... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strdup... yes
checking for strerror... yes
checking for strtol... yes
checking for strtoul... yes
checking whether our compiler supports __func__... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating man/Makefile
config.status: creating man/uncrustify.1
config.status: creating src/uncrustify_version.h
config.status: creating src/config.h
config.status: executing depfiles commands
[ 75%] No build step for 'external_uncrustify'
[ 87%] Performing install step for 'external_uncrustify'
Making install in src
Rebuilding token_names.h
sh: 0: Can't open ../make_token_names.sh
make[4]: *** [token_names.h] Error 127
make[3]: *** [install-recursive] Error 1
make[2]: *** [external_uncrustify-prefix/src/external_uncrustify-stamp/external_uncrustify-install] Error 2
make[1]: *** [CMakeFiles/external_uncrustify.dir/all] Error 2
make: *** [all] Error 2
```

I believe I'm using the latest:

```
$ cd src/ament/uncrustify && git log -1; cd -
commit b9cd1dc9b73eab61904491bf2a565622b34e5cc4
Author: Dirk Thomas <dthomas@osrfoundation.org>
Date:   Tue Apr 14 15:34:49 2015 -0700

    add package.xml and CMakeLists.txt
```

And I've completely cleared my `build` and `install` spaces before.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,12 +5,31 @@ project(uncrustify)
 include(ExternalProject)
 
 if(NOT WIN32)
+  # Make the destination for the source code.
+  set(uncrustify_source_dir "${CMAKE_CURRENT_BINARY_DIR}/uncrustify_src")
+  file(MAKE_DIRECTORY "${uncrustify_source_dir}")
+  # Touch a file there during configure time to prevent ExternalProject_Add from complaining about
+  # SOURCE_DIR being an empty folder.
+  execute_process(COMMAND
+    ${CMAKE_COMMAND} -E touch "${uncrustify_source_dir}/fool_external_project")
+  # Add a standard autotools external project, using the build space version of the source.
   ExternalProject_Add(
     external_uncrustify
-    SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}"
-    CONFIGURE_COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/configure"
+    SOURCE_DIR "${uncrustify_source_dir}"
+    CONFIGURE_COMMAND "${uncrustify_source_dir}/configure"
       "--prefix" "${CMAKE_CURRENT_BINARY_DIR}"
     BUILD_COMMAND "${MAKE}"
+  )
+  # Always copy the source of uncrustify to the build space between the download and build steps.
+  # This prevents problems that occur when the source space is a symlinked directory.
+  # See: https://github.com/ament/uncrustify/issues/1
+  ExternalProject_Add_Step(external_uncrustify copy_source
+    DEPENDEES download
+    DEPENDERS build
+    COMMENT "Copy source to ${uncrustify_source_dir} to prevent problems with symlink."
+    ALWAYS 1  # Always do this otherwise changing the source would not change the built binary.
+    COMMAND
+      ${CMAKE_COMMAND} -E copy_directory ${CMAKE_SOURCE_DIR}/ ${uncrustify_source_dir}
   )
 
   install(
@@ -33,7 +52,9 @@ else()
   add_custom_command(
     OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/win32/${build_type}/Uncrustify.exe"
     COMMAND "devenv" "/upgrade" "${CMAKE_CURRENT_SOURCE_DIR}/win32/Uncrustify_Win32_2011.sln"
-    COMMAND "msbuild" "${CMAKE_CURRENT_SOURCE_DIR}/win32/Uncrustify_Win32_2011.sln" "/p:Configuration=${build_type}" "/p:Platform=Win32"
+    COMMAND
+      "msbuild" "${CMAKE_CURRENT_SOURCE_DIR}/win32/Uncrustify_Win32_2011.sln"
+      "/p:Configuration=${build_type}" "/p:Platform=Win32"
     WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
     DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/src/uncrustify_version.h"
   )

